src/html-purify.js

```diff
@@ -70,30 +82,49 @@ See the accompanying LICENSE file for terms.
             idx = parser.getCurrentTagIndex();
             tagName = parser.getCurrentTag(idx);
 
-            if (contains(this.tagsWhitelist, tagName)) {
+            if (arrayLastIndexOf(this.tagsWhitelist, tagName) !== -1) {
 
                 if (idx) {
-                    if (this.config.enableTagBalancing) {
-                        // add closing tags for any opened ones before closing the current one
-                        while((openedTag = this.openedTags.pop()) && openedTag !== tagName) {
-                            this.output += '</' + openedTag + '>';
-                        }
-                        // openedTag is undefined if tagName is never found in all openedTags, no output needed
-                        if (openedTag) {
-                            this.output += '</' + openedTag + '>';
+                    if (tagBalance.enabled && !optionalElements[tagName]) {
+                        // relaxed tag balancing, accept it as long as the tag exists in the stack
```

Comment:
A document was prepared and shared internally on Aug 13. In case you missed it, please search "Untrusted input should be self-contained by tag balancing" in your mailbox.